### PR TITLE
comment stage + test

### LINKED ISSUE #667: Request Form: implement Comment Stage revision button
We should allow the PCs to enable/disable comments.

### PR DESCRIPTION
**Pull request by mohituniyal**
Fixes #667

### CODE REVIEW

File: openreview/venue_request/venue_request.py [hunk 1]
```diff
@@ -185,6 +185,81 @@ def setup_review_stage(self):
...
+                'description': 'Should the reviewers who have not submitted their review be able to make official comments? Default is "No, unsubmitted reviewers should not be able to post official comments"',
+                'value-radio': [
+                    'Yes, unsubmitted reviewers should be able to post official comments',
+                    'No, unsubmitted reviewers should not be able to post official comments'
+                ],
+                'required': True,
+                'default': 'No, unsubmitted reviewers should not be able to post official comments',
+                'order': 28
+            },
+            'allow_reader_selection': {
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:206]: I would prefer to remove this option, and always have the readers selectable.

File: openreview/venue_request/venue_request.py [hunk 2]
```diff
@@ -185,6 +185,81 @@ def setup_review_stage(self):
...
+                'description': 'Should the comment author be allowed to select comment readers from a list? Default is "No, comment authors should not be able to select comment readers"',
+                'value-radio': [
+                    'Yes, comment authors should be able to select comment readers',
+                    'No, comment authors should not be able to select comment readers'
+                ],
+                'required': True,
+                'default': 'No, comment authors should not be able to select comment readers',
+                'order': 29
+            },
+            'email_program_chairs_about_official_reviews': {
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:216]: what does it mean? official_reviews or official_comments?

File: openreview/venue_request/venue_request.py [hunk 3]
```diff
@@ -185,6 +185,81 @@ def setup_review_stage(self):
...
+                'description': 'Should the PCs receive an email for each official comment made in the venue? Default is "No, do not email PCs for each official comment in the venue"',
+                'value-radio': [
+                    'Yes, email PCs for each official comment made in the venue',
+                    'No, do not email PCs for each official comment made in the venue'
+                ],
+                'required': True,
+                'default': 'No, do not email PCs for each official comment made in the venue',
+                'order': 30
+            },
+            'allow_authors_to_comment': {
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:226]: I think `allow_authors_to_comment` and `allow_unsubmitted_reviewers_to_comment` should be together in a single field where the users can multi-select the participants of the discussions:

- public
- authors
- all reviewers
- assign reviewers
- submitted reviewers
- area chairs
- program chairs

I think if public is selected then we can create the public comments invitations. I'm not totally convinced about this, thoughts @mohituniyal @celestemartinez ?

File: openreview/venue_request/venue_request.py [hunk 4]
```diff
@@ -185,6 +185,81 @@ def setup_review_stage(self):
...
+                    'No, do not allow submission authors to post official comments on their own papers'
+                ],
+                'required': True,
+                'default': 'Yes, allow submission authors to post official comments on their own papers',
+                'order': 31
+            }
+        }
+
+        return self.venue_request.client.post_invitation(openreview.Invitation(
+            id='{}/-/Official_Comment_Stage'.format(self.venue_request.support_group.id),
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:239]: can we call it `Comment_Stage`?

File: openreview/venue_request/venue_request.py [hunk 5]
```diff
@@ -185,6 +185,81 @@ def setup_review_stage(self):
             }
         ))
 
+    def setup_official_comment_stage(self):
+
+        official_comment_stage_content = {
+            'commentary_start_date': {
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:191]: don't we need an end date?

File: openreview/venue_request/venue_request.py (hunk 2 above)
Comment (mohituniyal) [reply openreview/venue_request/venue_request.py:216]: You are right, it's official comments, sorry.

File: openreview/venue_request/venue_request.py (hunk 5 above)
Comment (mohituniyal) [reply openreview/venue_request/venue_request.py:191]: @melisabok I have included this in the latest changes. The parameter is called `end_date` as it corresponds to the expiry date for comment invitations. I have applied start date and end date to both public and official comment invitations. Please take a look.

File: openreview/venue_request/venue_request.py (hunk 3 above)
Comment (mohituniyal) [reply openreview/venue_request/venue_request.py]: Changes have been made for all other comments. About this one, I like the idea of aggregating all the separate readership questions into 1. The field name and description would be very important though.

File: openreview/venue_request/venue_request.py [hunk 6]
```diff
@@ -185,6 +185,76 @@ def setup_review_stage(self):
...
+                'description': 'When does official and/or public commentary begin? Please use the following format: YYYY/MM/DD HH:MM (e.g. 2019/01/31 23:59)',
+                'value-regex': r'^[0-9]{4}\/([1-9]|0[1-9]|1[0-2])\/([1-9]|0[1-9]|[1-2][0-9]|3[0-1])(\s+)?((2[0-3]|[01][0-9]|[0-9]):[0-5][0-9])?(\s+)?$',
+                'order': 27
+            },
+            'commentary_end_date': {
+                'description': 'When does official and/or public commentary end? Please use the following format: YYYY/MM/DD HH:MM (e.g. 2019/01/31 23:59)',
+                'value-regex': r'^[0-9]{4}\/([1-9]|0[1-9]|1[0-2])\/([1-9]|0[1-9]|[1-2][0-9]|3[0-1])(\s+)?((2[0-3]|[01][0-9]|[0-9]):[0-5][0-9])?(\s+)?$',
+                'order': 28
+            },
+            'allow_unsubmitted_reviewers_to_comment': {
```
Comment (melisabok) [inline openreview/venue_request/venue_request.py:201]: merge these two fields in a single one:

participants:

- Program Chairs
- All Area Chairs
- Paper Area Chairs
- All Reviewers
- Paper Reviewers
- Submitted Reviewers
- Authors

Comment (melisabok) [reply openreview/venue_request/venue_request.py:201]: this means that the selected values will be invitees and part of the readers selection.

File: openreview/venue_request/venue_request.py (hunk 3 above)
Comment (melisabok) [reply openreview/venue_request/venue_request.py]: this change that Mohit is saying is not implemented, I can not find it. @celestemartinez can you work on this?

**Comment by celestemartinez** [comment]
I was thinking of removing 'Public Commentary' from the venue request because whether we enable public comments or not depends on the 'Participants' field. I am ok with leaving it there as well. What do you think @melisabok?

**Comment by melisabok** [comment]
Ohhh yeah, I agree, we should remove it from the request form invitation. They can use the CommentStage to define the policy.

File: openreview/conference/helpers.py [hunk 7]
```diff
@@ -305,3 +305,43 @@ def get_submission_revision_stage(client, request_forum):
...
+    if commentary_end_date:
+        try:
+            commentary_end_date = datetime.datetime.strptime(commentary_end_date, '%Y/%m/%d %H:%M')
+        except ValueError:
+            commentary_end_date = datetime.datetime.strptime(commentary_end_date, '%Y/%m/%d')
+    else:
+        commentary_end_date = None
+
+    allow_public_comments = 'Public' in request_forum.content.get('participants', '')
+    anonymous = allow_public_comments and 'non-anonymously' not in request_forum.content.get('participants', '')
```
Comment (melisabok) [inline openreview/conference/helpers.py:330]: may be we should have two options like:

`Non anonymous public`
`anonymous public`
?

Comment (melisabok) [reply openreview/conference/helpers.py:330]: Never mind, I saw the two options :)

**Review by melisabok — APPROVED** [review]
This is great! 🙆‍♀️

**Pull request merged**